skimage/measure/mc_meta/visual_test.py

```diff
@@ -3,70 +3,118 @@
+    m2 = vv.mesh(np.fliplr(vertices2), faces2)
+    a1.camera = a2.camera
+
+    # visvis uses right-hand rule, gradient_direction param uses left-hand rule
+    m1.cullFaces = m2.cullFaces = 'front'  # None, front or back
 
-if SELECT == 1:
-    # Medical data
-    vol = vv.volread('stent')
-    isovalue = 800
+    vv.use().Run()
 
-elif SELECT == 2:
-    # Blocky data
-    vol = vv.aVolume(20, 128) # Different every time
-    isovalue = 0.2
 
-elif SELECT == 3:
-    # Generate two donuts using a formula by Thomas Lewiner
+def donuts():
+    """Return volume of two donuts based on a formula by Thomas Lewiner."""
     n = 48
-    a, b = 2.5 / n, -1.25
-    isovalue = 0.0
-    #
+    a = 2.5 / n * 8.0
+    b = -1.25 * 8.0
+    c = 16.0 - 1.85 * 1.85
+    d = 64.0
     vol = np.empty((n, n, n), 'float32')
-    for iz in range(vol.shape[0]):
-        for iy in range(vol.shape[1]):
-            for ix in range(vol.shape[2]):
-                z, y, x = float(iz)*a+b, float(iy)*a+b, float(ix)*a+b
-                vol[iz, iy, ix] = ( (
-                    (8*x)**2 + (8*y-2)**2 + (8*z)**2 + 16 - 1.85*1.85 ) * ( (8*x)**2 +
-                    (8*y-2)**2 + (8*z)**2 + 16 - 1.85*1.85 ) - 64 * ( (8*x)**2 + (8*y-2)**2 )
-                    ) * ( ( (8*x)**2 + ((8*y-2)+4)*((8*y-2)+4) + (8*z)**2 + 16 - 1.85*1.85 )
-                    * ( (8*x)**2 + ((8*y-2)+4)*((8*y-2)+4) + (8*z)**2 + 16 - 1.85*1.85 ) -
-                    64 * ( ((8*y-2)+4)*((8*y-2)+4) + (8*z)**2
-                    ) ) + 1025
-    # Uncommenting the line below will yield different results for classic MC
-    #vol = -vol
-
-elif SELECT == 4:
-    vol = ellipsoid(4, 3, 2, levelset=True)
-    isovalue = 0
-
-# Get surface meshes
-t0 = time.time()
-vertices1, faces1, _, _ = marching_cubes_lewiner(vol, isovalue, gradient_direction=gradient_dir, use_classic=False)
-print('finding surface lewiner took %1.0f ms' % (1000*(time.time()-t0)) )
-
-t0 = time.time()
-vertices2, faces2 = marching_cubes_classic(vol, isovalue, gradient_direction=gradient_dir)
-print('finding surface classic took %1.0f ms' % (1000*(time.time()-t0)) )
-
-# Show
-vv.figure(1); vv.clf()
-a1 = vv.subplot(121); m1 = vv.mesh(np.fliplr(vertices1), faces1)
-a2 = vv.subplot(122); m2 = vv.mesh(np.fliplr(vertices2), faces2)
-a1.camera = a2.camera
-
-# visvis uses right-hand rule, gradient_direction param uses left-hand rule
-m1.cullFaces = m2.cullFaces = 'front'  # None, front or back
-
-vv.use().Run()
+    for iz in range(n):
+        z = iz * a + b
+        z *= z
+        zc = z + c
+        for iy in range(n):
+            y = iy * a + b
+            y1 = y - 2
+            y1 *= y1
+            y2 = y + 2
+            y2 *= y2
+            for ix in range(n):
+                x = ix * a + b
+                x *= x
+                x1 = x + y1 + zc
+                x1 *= x1
+                x2 = x + y2 + zc
+                x2 *= x2
+                vol[iz, iy, ix] = ((x1 - d * (x + y1)) * (x2 - d * (z + y2)))

```

Comment:
This looks eminently vectorisable...?

Reply:
I don't think this is relevant since this function is not part of the skimage "library". It can easily be speed up as is by orders of magnitude with numba.jit or cythonizing with prange, Vectorizing would probably make it less readable and use more memory.

Reply:
I agree that it'll use more memory, but I disagree that it'll be less readable. It would presumably read like an actual formula, which right now I can't for the life of me figure out based on this code. But on the other hand this was even more true in the previous version of the code! So, as mentioned before, this is fine. =)

Reply:
Can we at least agree that constructs such as:
```python
z = iz * a + b
z *= z
``` 
should be replaced by
```python
z = (iz * a + b) ** 2
```

There are many of them in here, and github's suggestion feature doesn't see appropriate.

Reply:
I guess not, it makes things slower. 40 ms as written, and 47 ms with the `(  ) ** 2` suggestion. 

Using fully vectorized code, it goes down to 2ms

```
def donuts():
    """Return volume of two donuts based on a formula by Thomas Lewiner."""
    n = 48
    a = 2.5 / n * 8.0
    b = -1.25 * 8.0
    c = 16.0 - 1.85 * 1.85
    d = 64.0
    
    i = np.arange(n, dtype=int)
    ia_plus_b = i * a + b
    ia_plus_b_square = ia_plus_b ** 2
    z = ia_plus_b_square[:, np.newaxis, np.newaxis]
    zc = z + c
    
    y1 = ((ia_plus_b - 2) ** 2)[np.newaxis, :, np.newaxis]
    y2 = ((ia_plus_b + 2) ** 2)[np.newaxis, :, np.newaxis]
    
    x = ia_plus_b_square[np.newaxis, np.newaxis, :]
    x1 = (ia_plus_b_square + y1 + zc) ** 2
    x2 = (ia_plus_b_square + y2 + zc) ** 2
    
    return ((x1 - d * (x + y1)) * (x2 - d * (z + y2))) + 1025
```

Reply:
> I guess not, it makes things slower. 40 ms as written, and 47 ms with the `( ) ** 2` suggestion.

Well, in this case, readability trumps speed, or we wouldn't be doing such a loop in Python.

But I agree that the vectorised version is better.